servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/P2CSelector.java

```diff
@@ -62,29 +62,19 @@ public Single<C> selectConnection(
                 throw new AssertionError("Selector for " + getTargetResource() +
                         " received an empty host set");
             case 1:
-                Host<ResolvedAddress, C> host = hosts.get(0);
-                if (!forceNewConnectionAndReserve) {
-                    C connection = host.pickConnection(selector, context);
-                    if (connection != null) {
-                        return succeeded(connection);
-                    }
-                }
-                // Either we require a new connection or there wasn't one already established so
-                // try to make a new one if the host is healthy. If it's not healthy, we fail
-                // and let the higher level retries decide what to do.
-                if (!host.isActiveAndHealthy()) {
-                    return noActiveHosts(hosts);
-                }
-                return host.newConnection(selector, forceNewConnectionAndReserve, context);
+                // There is only a single host, so we don't need to do any of the looping or comparison logic.
+                Single<C> connection = selectFromHost(hosts.get(0), selector, forceNewConnectionAndReserve, context);
+                return connection == null ? noActiveHosts(hosts) : connection;
             default:
                 return p2c(size, hosts, getRandom(), selector, forceNewConnectionAndReserve, context);
         }
     }
 
-    @Nullable
     private Single<C> p2c(int size, List<Host<ResolvedAddress, C>> hosts, Random random, Predicate<C> selector,
                           boolean forceNewConnectionAndReserve, @Nullable ContextMap contextMap) {
-        for (int j = maxEffort; j > 0; j--) {
+        // If there are only two hosts we only try once since there is no chance we'll select different hosts
+        // on further iterations.
+        for (int j = hosts.size() == 2 ? 1 : maxEffort; j > 0; j--) {
```

Comment:
Will it make sense to do smth like: `max(1, min(maxEffort, hosts.size() - 1))`?

Reply:
I'm personally inclined not to. Since selection is random there isn't a guarantee that we won't make effectively the same selections a few times so I'd like to keep it simple and just exhaust the `maxEffort`. That ways we don't have to try and guess how many times we've attempted a selection based on the host set which has a dynamic size.